src/audio/src/src_hifi4.c

```diff
@@ -169,139 +188,145 @@ static inline void fir_filter(ae_f32 *rp, const void *cp, ae_f32 *wp0,
+	/* Move data pointer back by one sample to start from right
+	 * channel sample. Discard read value p0.
+	 */
+	dp = (ae_f32x2 *)rp;
+	AE_L32_XC(f0, (ae_f32 *)dp, -sizeof(ae_f32));
+
+	/* Reset coefficient pointer and clear accumulator */
+	coefp = (ae_int32x2 *)cp;
+	coef_align = AE_LA64_PP(coefp);
+
+	/* Compute FIR filter for current channel with four
+	 * taps per every loop iteration.  Two coefficients
+	 * are loaded simultaneously. Data is read
+	 * from interleaved buffer with stride of channels
+	 * count.
+	 */
+	for (i = 0; i < taps_div_4; i++) {
+		/* Load four coefficients */
+		AE_LA32X2_IP(coef10, coef_align, coefp);
+		AE_LA32X2_IP(coef32, coef_align, coefp);
+
+		/* Load two data samples from two channels. Note: Due to
+		 * polyphase array data start shift for sub-filters can't
+		 * use 128 bits load due to align requirement.
 		 */
-		dp = (ae_f24x2 *)rp;
-		AE_L32F24_XC(d0, (ae_f24 *)dp, -sizeof(ae_f24));
+		AE_L32X2_XC(f0, dp, inc); /* f0.h is left0, f0.l is right0 */
+		AE_L32X2_XC(f1, dp, inc); /* f1.h is left1, f1.l is right1 */
 
-		/* Reset coefficient pointer and clear accumulator */
-		coefp = (ae_f24x2 *)cp;
-		a0 = AE_ZERO64();
-		a1 = AE_ZERO64();
-
-		/* Compute FIR filter for current channel with four
-		 * taps per every loop iteration.  Two coefficients
-		 * are loaded simultaneously. Data is read
-		 * from interleaved buffer with stride of channels
-		 * count.
+		/* a0 (left)  += left10.h  (left1)  * coef10.h (coef2)
+		 *            += left10.l  (left0)  * coef10.l (coef1)
+		 * a1 (right) += right10.h (right1) * coef10.h (coef2)
+		 *            += right10.l (right0) * coef10.l (coef1)
 		 */
-		for (i = 0; i < taps_div_4; i++) {
-			/* Load two coefficients. Coef2_h contains tap *coefp
-			 * and coef2_l contains the next tap.
-			 */
-			/* TODO: Ensure coefficients are 64 bits aligned */
-			AE_L32X2F24_IP(coef2, coefp, sizeof(ae_f24x2));
-
-			/* Load two data samples from two channels */
-			AE_L32X2F24_XC(d0, dp, inc); /* r0, l0 */
-			AE_L32X2F24_XC(d1, dp, inc); /* r1, l1 */
-
-			/* Select to data2 two successive left channel samples
-			 * from d0 and d1, multiply-add and accumulate to a0.
-			 * Select to data2 two successive right channel samples
-			 * from d0 and d1, multiply-add and accumulate to a1.
-			 * data2_h * coef2_h + data2_l * coef2_l. The Q1.31
-			 * data and Q1.15 coefficients are used as 24 bits as
-			 * Q1.23 values.
-			 */
-			data2 = AE_SELP24_LL(d0, d1);
-			AE_MULAAFP24S_HH_LL(a0, data2, coef2);
-			data2 = AE_SELP24_HH(d0, d1);
-			AE_MULAAFP24S_HH_LL(a1, data2, coef2);
-
-			/* Repeat for next two taps */
-			AE_L32X2F24_IP(coef2, coefp, sizeof(ae_f24x2));
-			AE_L32X2F24_XC(d0, dp, inc); /* r2, l2 */
-			AE_L32X2F24_XC(d1, dp, inc); /* r3, l3 */
-			data2 = AE_SELP24_LL(d0, d1);
-			AE_MULAAFP24S_HH_LL(a0, data2, coef2);
-			data2 = AE_SELP24_HH(d0, d1);
-			AE_MULAAFP24S_HH_LL(a1, data2, coef2);
-		}
+		left10 = AE_SEL32_HH(f0, f1);
```

Comment:
Save one AE register variable and use just sample10 instead of left10 and right10.